homeassistant/components/mqtt/__init__.py

```diff
@@ -530,3 +556,50 @@ async def async_remove_config_entry_device(
 
     await device_automation.async_removed_from_device(hass, device_entry.id)
     return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload MQTT dump and publish service when the config entry is unloaded."""
+    # Unload publish and dump services.
+    hass.services.async_remove(
+        DOMAIN,
+        SERVICE_PUBLISH,
+    )
+    hass.services.async_remove(
+        DOMAIN,
+        SERVICE_DUMP,
+    )
+
+    # Unload the MQTT platforms
+    async with hass.data[DATA_CONFIG_ENTRY_LOCK]:
+        # Stop the discovery
+        await discovery.async_stop(hass)
+        connection: MQTT = hass.data[DATA_MQTT]
+        # Unload the platforms
+        await asyncio.gather(
+            *(
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in PLATFORMS
+            )
+        )
+        # Unsubscribe discovery dispatchers
+        while discovery_dispatchers := hass.data.setdefault(
+            DATA_MQTT_DISCOVERY_DISPATCHERS, []
+        ):
+            discovery_dispatchers.pop(0)()
+        hass.data[CONFIG_ENTRY_IS_SETUP] = set()
+        # Cleanup listeners
+        connection.cleanup()
+
+    # Trigger reload manual MQTT items at entry setup
+    hass.data[DATA_MQTT_RELOAD_NEEDED] = True
+    if (mqtt_entry := mqtt_config_entry_enabled(hass)) is False:
```

Comment:
The variable name is confusing, please rename it to `mqtt_entry_enabled` or `entry_enabled` to not suggest it holds a config entry.
Also, it must be possible to end up here and have `mqtt_config_entry_enabled` return `None`?